Add support for linear solvers : Petsc and Hypre

latug0 opened the pull request:
Add support for linear solvers defined within MFEM : Petsc and Hypre.
Add also a support for MED files if dedicated MFEM version is used.
The support for several preconditionners are added too.
Developed by T. Helfer, H.copin, G. Latu.

latug0 commented:
I wonder why Mult is called two times in NonLinearPbImplBase:solve whever PETSc is used ??
```
#ifdef MFEM_USE_PETSC
    if (usePETSc()) {
      this->petsc_solver->Mult(zero, this->u1);
      return this->petsc_solver->GetConverged();
    }
#endif /* MFEM_USE_PETSC */
    this->solver->Mult(this->u0, this->u1);
    return this->solver->GetConverged();
```

thelfer commented:
Hi Guillaume,
I don't see in which case `Mult` could be called twice. The code sounds good to me.
Best,
Thomas

latug0 commented:
> I don't see in which case `Mult` could be called twice. The code sounds good to me.

Right. I added some comments about the "zero" argument for the PETSc call. 
And I changed a little bit the `#ifdef` in order to have a clearer conditionnal IMHO.
Regards.
G.
